Sources/PriorityQueueModule/PriorityQueue.swift

```diff
@@ -0,0 +1,490 @@
+  @inline(__always)
+  internal func _minMaxHeapIsMinLevel(_ count: Int) -> Bool {
+    precondition(count > 0)
+
+    return count._binaryLogarithm() & 0b1 == 0
+  }
+
+  /// Swaps the elements in the heap at the given indices.
+  @inline(__always)
+  private mutating func _swapAt(_ i: Int, _ j: Int) {
+    let tmp = storage[i]
+    storage[i] = storage[j]
+    storage[j] = tmp
+  }
+
+  /// Returns the parent index of the given `index`
+  /// or `nil` if the index has no parent (i.e. `index == 0`).
+  @inline(__always)
+  private func _parentIndex(of index: Int) -> Int? {
+    guard index > 0 else {
+      return nil
+    }
+
+    return (index - 1) / 2
+  }
+
+  /// Returns the grandparent index of the given `index`
+  /// or `nil` if the index has no grandparent.
+  @inline(__always)
+  private func _grandparentIndex(of index: Int) -> Int? {
+    guard index > 2 else {
+      return nil
+    }
+
+    return (index - 3) / 4
+  }
+
+  /// Returns the first child index of the given `index`
+  /// or `nil` if the index has no children.
+  @inline(__always)
+  private func _leftChildIndex(of index: Int) -> Int? {
+    let childIdx = index * 2 + 1
+    guard childIdx < storage.count else {
+      return nil
+    }
+
+    return childIdx
+  }
+
+  /// Returns the right child index of the given `index`
+  /// or `nil` if the index has no right child.
+  @inline(__always)
+  private func _rightChildIndex(of index: Int) -> Int? {
+    let childIdx = index * 2 + 2
+    guard childIdx < storage.count else {
+      return nil
+    }
+
+    return childIdx
+  }
+
+  /// Returns the first grandchild index of the given `index`
+  /// or `nil` if the index has no grandchildren.
+  @inline(__always)
+  private func _firstGrandchildIndex(of index: Int) -> Int? {
+    let grandchildIdx = index * 4 + 3
+    guard grandchildIdx < storage.count else {
+      return nil
+    }
+
+    return grandchildIdx
+  }
+
+  /// Returns the last valid grandchild index of the given `index`
+  /// or `nil` if the index has no grandchildren.
+  ///
+  /// In cases where the given index only has one grandchild, the index
+  /// returned by this function is the same as that returned by
+  /// `_firstGrandchildIndex`.
+  @inline(__always)
+  private func _lastGrandchildIndex(of index: Int) -> Int? {
+    guard _firstGrandchildIndex(of: index) != nil else {
+      // There are no grandchildren of the node at `index`
+      return nil
+    }
+
+    return Swift.min(index * 4 + 6, storage.count - 1)
+  }
+}
+
+// MARK: -
+
+extension PriorityQueue {
+  /// Initializes a queue from a collection.
+  ///
+  /// Utilizes [Floyd's linear-time heap construction algorithm](https://en.wikipedia.org/wiki/Heapsort#Floyd's_heap_construction).
+  public init<C: Collection>(_ collection: C) where C.Element == Element {
+    storage = Array(collection)
+
+    for idx in (0..<(storage.count / 2)).reversed() {
+      _trickleDown(startingAt: idx)
+    }
+  }
```

Comment:
This linear-time initializer is awesome!

I think the algorithm over `Collection` is redundant with the more general algorithm over `Sequence` below.

Reply:
Yeah, I ran into some issues with these (though I don't recall what they were now). I'll try to remove the `Collection` initializer and re-add the `ExpressibleByArrayLiteral` conformance.